### PR TITLE
Fix lodash scripts

### PR DESCRIPTION
**Pull request by whitphx**
I couldn't maintain [this PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/42220) and it was closed, then now I recreate here again.

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:
(None of them)

I fixed the development scripts and configs inside `types/lodash` and update the auto-generated modules related to them.
This is based on https://github.com/DefinitelyTyped/DefinitelyTyped/issues/41857#issuecomment-580417935

### CODE REVIEW

File: types/lodash/tslint.json
Comment (sandersn) [inline types/lodash/tslint.json:9]: why is this needed?

File: types/lodash/scripts/package.json
Comment (sandersn) [inline types/lodash/scripts/package.json:6]: seems like this should change to be `"modules": "ts-node generate-modules"`, right?

File: types/lodash/scripts/generate-all.sh
Comment (sandersn) [inline types/lodash/scripts/generate-all.sh:4]: in that case maybe we should just delete it.

File: types/lodash/scripts/generate-fp.ts
Comment (sandersn) [inline types/lodash/scripts/generate-fp.ts:132]: is the tsconfig generated somewhere else now or is it not needed?

**Review by sandersn — CHANGES_REQUESTED** [review]
A few questions. And since I haven't spent much time looking at the scripts, @aj-r could you review this too if you have time?

**Comment by sandersn** [comment]
Actually, it looks like @aj-r reviewed the previous version of this. @tuttieee I'll merge this after you answer the questions I had.

File: types/lodash/scripts/generate-fp.ts
Comment (tuttieee) [reply types/lodash/scripts/generate-fp.ts:132]: > is the tsconfig generated somewhere else

Yes. Here: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/bf8afd29658567c0c5cd447b3e508a9ec9dcaaf3/types/lodash/scripts/generate-modules.ts#L21

---

Apart from that, I was not sure if it's really ok to remove the here, but thought there would be no drawbacks in practice and decided to do so.
The details are below:

What the code here does is to add `fp/*` files to the "files" section in `tsconfig.json`, and if the code remains, this script `generate-fp.ts` overwrites `tsconfig.json` with the content like the following
```json
{
    "compilerOptions": {
        "module": "commonjs",
        "lib": [
            "es6"
        ],
        "noImplicitAny": true,
        "noImplicitThis": true,
        "strictNullChecks": true,
        "strictFunctionTypes": true,
        "baseUrl": "../",
        "typeRoots": [
            "../"
        ],
        "types": [],
        "noEmit": true,
        "forceConsistentCasingInFileNames": true
    },
    "files": [
        "index.d.ts",
        "lodash-tests.ts",
        "fp/add.d.ts",
        "fp/after.d.ts",
...(many lines)...
        "fp/placeholder.d.ts"
    ]
}
```

However, `tsconfig.json` actually committed to this repo does not include these lines ( https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/lodash/tsconfig.json#L20-L23 ),
which means the authors have never used `generate-fp.ts` for a long time or they have ignored the generated `tsconfig.json` to commit.
And when I removed the code here like this PR, the generated content of `tsconfig.ts` exactly matches the current committed one - this modified `generate-fp.ts` outputs `tsconfig.json` as-is.
That's why I thought there would be no drawbacks *in practice* if I delete the code here and decided to do so.

File: types/lodash/scripts/generate-all.sh
Comment (tuttieee) [reply types/lodash/scripts/generate-all.sh:4]: Right.

File: types/lodash/scripts/package.json
Comment (tuttieee) [reply types/lodash/scripts/package.json:6]: That's a matter of design decision.

Which do you think is the best?

### Option1 (current):
```
  "scripts": {
    "generate": "ts-node generate-modules && ts-node generate-fp"
  },
```

### Option2:
```
  "scripts": {
    "modules": "ts-node generate-modules",
    "fp": "ts-node generate-fp"
  },
```

### Option 2'
```
  "scripts": {
    "all": "npm run modules && npm run fp",
    "modules": "ts-node generate-modules",
    "fp": "ts-node generate-fp"
  },
```

## Things to consider - from my understanding
* Authors have to run both `generate-modules.ts` and `generate-fp.ts` to generate both `<module-name>.d.ts` and `fp/*.d.ts`
* The instructions are https://github.com/DefinitelyTyped/DefinitelyTyped/blob/bf8afd29658567c0c5cd447b3e508a9ec9dcaaf3/types/lodash/readme.md#before-creating-a-pr
  * This is one of the reasons why I chose option1 - the README already instructs authors to run `npm run generate` and I thought it should contain all necessary scripts.
  * But of course it's also OK to rewrite README to instruct like "Re-generate the files by opening a terminal in the scripts directory and running `npm run all` for example if we choose option 2' above.

File: types/lodash/tslint.json
Comment (tuttieee) [reply types/lodash/tslint.json:9]: Please see https://github.com/DefinitelyTyped/DefinitelyTyped/pull/42220#discussion_r376771490

File: types/lodash/scripts/generate-all.sh
Comment (tuttieee) [reply types/lodash/scripts/generate-all.sh:4]: I was not sure if it was truly OK to delete the file because I haven't been fully familiar with the development workflow of @types/lodash and while @aj-r taught me about `generate-all.sh` in https://github.com/DefinitelyTyped/DefinitelyTyped/issues/41857#issuecomment-580417935, it was not 100% decisive...
but yes, I should have deleted it.
It's one reason we are using VCS :)

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## lodash.multiply ([<kbd>unpkg</kbd>](https://unpkg.com/browse/lodash.multiply@latest/))
was missing the following properties:
1. apply
1. call
1. bind
1. toString
1. length

<details>
<summary>as well as these 3 other properties...</summary>
<p>arguments, caller, name</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against dccb37bfcb442e392962dc2bcf118cdbeb48d9f6
</p>

**Comment by tuttieee** [comment]
This PR is not responsible for the topic posted by dangerJS above.

File: types/lodash/scripts/generate-fp.ts
Comment (aj-r) [reply types/lodash/scripts/generate-fp.ts:132]: Looks like those files were removed from tsconfig relatively recently (Nov 2019, #40676) so it is correct to remove this code. Its purpose was to add all of the "fp/" files but it looks like that is no longer necessary.

However, it looks like `index.d.ts` does not reference `fp.d.ts`, so I wonder if we should add `fp.d.ts` to the files list.

File: types/lodash/scripts/generate-modules.ts
Comment (aj-r) [inline types/lodash/scripts/generate-modules.ts:22]: Maybe we should stop auto-generating this file? If its contents are not based on the list of modules, then I'm inclined to let people edit it manually instead of auto-generating it.

**Review by aj-r — CHANGES_REQUESTED** [review]

File: types/lodash/scripts/generate-modules.ts
Comment (tuttieee) [reply types/lodash/scripts/generate-modules.ts:22]: I agree, but am not familiar with the history and why it's generated by the script, then not sure if it's OK...

File: types/lodash/scripts/package.json
Comment (sandersn) [reply types/lodash/scripts/package.json:6]: I think your choice is good -- it's backward compatible and a small-ish change.

**Review by sandersn — APPROVED** [review]
All my comments are addressed, but we should probably hold off on merging until we've decided what to do about @aj-r's question of whether to stop auto-generating tsconfigs.

**Comment by aj-r** [comment]
FYI there's another PR (#27224) which adds all of the fp file references back (probably because they ran the generate script). I'd like to know if these files really are needed... if they are, we should keep the auto-generate code. If not, I think we should stop auto-generating `tsconfig.json` altogether.

If we're not sure, then let's keep generating them and we can fix in a follow-up PR if needed (to avoid delaying this PR).

**Comment by tuttieee** [comment]
> I'd like to know if these files really are needed

Me too.

> If we're not sure, then let's keep generating them and we can fix in a follow-up PR if needed (to avoid delaying this PR).

I agree.
Then I will restore the generating script to keep it as-is.

**Comment by tuttieee** [comment]
@aj-r Thank you,

> I'd like to know if these files really are needed

Me too.

> If we're not sure, then let's keep generating them and we can fix in a follow-up PR if needed (to avoid delaying this PR).

I agree.
Then I will restore the generating script to keep it as-is.

**Comment by tuttieee** [comment]
mm, wait, I think the PR you mentioned (#27224) does not add `fp/*` files to `tsconfig.json`
and `tsconfig.json` in the master branch does not have such contents: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/lodash/tsconfig.json

**Comment by tuttieee** [comment]
Then, now I still think the current changes in this PR are good enough.

### `fp/*` items added to `files` field by `scripts/generate-fp.ts`
The code in `scripts/generate-fp.ts` that adds `fp/*` to `tsconfig.json` should be removed because the resultant `tsconfig.json`'s content does not match the current one.

In addition, since the current `tsconfig.json` in the master branch does not contain such fields, there will be no effects if we remove the unused generator.


### Auto-generation of `tsconfig.json`
The code in `scripts/generate-modules.ts` that generates `tsconfig.json` should be preserved because it also generates `../lodash.<module_name>/tsconfig.json`-s and ensures that these `tsconfig.json`-s have the same `compilerOptions` fields.

Please see that [this code](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/lodash/scripts/generate-modules.ts#L90) that generates `tsconfig.json` and [this code](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/lodash/scripts/generate-modules.ts#L101) that generates `../lodash.<module_name>/tsconfig.json` are using the same `compilerOptions()`

And because of this, the answer to [this discussion](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44486#discussion_r426269883) is now NO.
> Maybe we should stop auto-generating this file?

The auto-generation seems responsible to make all `tsconfig.json`-s `compilerOptions` synced.

---

Besides, I extracted the change that removes the code into a separate commit (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44486/commits/84b3501e5c6c17f9164ab3ac9bf0742db27286ff) so that we can revert only the change easily if it turned out necessary.
So now I have to ask you to preview again, sorry.

**Review by aj-r — APPROVED** [review]
Yes it looks like that PR reverted the change to tsconfog.json. Looks good.

**Comment by tuttieee** [comment]
Mmm, the CI seems to have already finished, but the status on GitHub is still "in progress".
Does anyone know how to fix it? Is the only way to recreate the commits and force push again?

**Comment by tuttieee** [comment]
Oh solved without requesting review again simply by `git commit --amend` and `git push -f`.

**Comment by tuttieee** [comment]
But it seems owners' approvals are needed again 😥
@aj-r @sandersn Can you approve again?

**Review by aj-r — APPROVED** [review]

**Review by sandersn — APPROVED** [review]

**Comment by tuttieee** [comment]
Ready to merge

**Pull request merged**
